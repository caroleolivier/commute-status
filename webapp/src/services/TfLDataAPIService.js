/* eslint-env browser */
/* eslint-disable class-methods-use-this */

import ArrivalTime from './ArrivalTime';

class TfLDataAPIService {
    fetchArrivals(stationId, directionId, lines) {
        const url = this._buildUrl(stationId, directionId, lines);
        return fetch(url)
            .then(resp => resp.json())
            .then(json => this._parseData(json));
    }

    _buildUrl(stationId, directionId, lines) {
        const linesStr = lines.join('%2C', lines);
        return `https://api.tfl.gov.uk/Line/${linesStr}/Arrivals/${stationId}?direction=${directionId}`;
    }

    _parseData(json) {
        return json.map((arrival) => {
            const lineName = arrival.lineName;
            const destinationName = arrival.destinationName;
            const expectedIn = arrival.timeToStation;
            return new ArrivalTime(lineName, destinationName, expectedIn);
        }).sort((a, b) => a.expectedSeconds - b.expectedSeconds);
    }

    // fetchArrivals() {
    //             const json = 
    //         [{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"1152176492","operationType":1,"vehicleId":"011","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":45,"currentLocation":"Between Tooting Bec and Tooting Broadway","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:05:28Z","timeToLive":"2017-10-09T08:05:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.307Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-506176823","operationType":1,"vehicleId":"012","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":705,"currentLocation":"At Stockwell Platform 3","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:16:28Z","timeToLive":"2017-10-09T08:16:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.338Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-2105692590","operationType":1,"vehicleId":"013","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1185,"currentLocation":"Between Borough and Elephant and Castle ","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:24:28Z","timeToLive":"2017-10-09T08:24:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.354Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"1554330431","operationType":1,"vehicleId":"014","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1485,"currentLocation":"At Bank Platform 3","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:29:28Z","timeToLive":"2017-10-09T08:29:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.369Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-104692584","operationType":1,"vehicleId":"015","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1605,"currentLocation":"At Old Street Platform 2","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:31:28Z","timeToLive":"2017-10-09T08:31:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.385Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"1151586668","operationType":1,"vehicleId":"041","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":825,"currentLocation":"Between Oval and Stockwell","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:18:28Z","timeToLive":"2017-10-09T08:18:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.338Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-506766647","operationType":1,"vehicleId":"042","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1005,"currentLocation":"At Kennington Platform 2","towards":"Morden via CX","expectedArrival":"2017-10-09T08:21:28Z","timeToLive":"2017-10-09T08:21:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.354Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-2106544558","operationType":1,"vehicleId":"043","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1605,"currentLocation":"Between Goodge Street and Tottenham Court Road","towards":"Morden via CX","expectedArrival":"2017-10-09T08:31:28Z","timeToLive":"2017-10-09T08:31:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.385Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"1553478463","operationType":1,"vehicleId":"044","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1725,"currentLocation":"Between Angel and Old Street","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:33:28Z","timeToLive":"2017-10-09T08:33:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.401Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"882351566","operationType":1,"vehicleId":"057","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":225,"currentLocation":"At Balham Platform 2","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:08:28Z","timeToLive":"2017-10-09T08:08:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.307Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"1553871679","operationType":1,"vehicleId":"064","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":585,"currentLocation":"Between Clapham North and Clapham Common","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:14:28Z","timeToLive":"2017-10-09T08:14:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.322Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-104364904","operationType":1,"vehicleId":"065","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1185,"currentLocation":"At Borough Platform 2","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:24:28Z","timeToLive":"2017-10-09T08:24:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.369Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"1151455597","operationType":1,"vehicleId":"161","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":465,"currentLocation":"Departed Clapham Common","towards":"Morden via CX","expectedArrival":"2017-10-09T08:12:28Z","timeToLive":"2017-10-09T08:12:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.322Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-506897674","operationType":1,"vehicleId":"162","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":945,"currentLocation":"At Kennington Platform 4","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:20:28Z","timeToLive":"2017-10-09T08:20:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.338Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}},{"$type":"Tfl.Api.Presentation.Entities.Prediction, Tfl.Api.Presentation.Entities","id":"-2106151341","operationType":1,"vehicleId":"163","naptanId":"940GZZLUTBY","stationName":"Tooting Broadway Underground Station","lineId":"northern","lineName":"Northern","platformName":"Southbound - Platform 2","direction":"inbound","bearing":"","destinationNaptanId":"940GZZLUMDN","destinationName":"Morden Underground Station","timestamp":"2017-10-09T08:04:43Z","timeToStation":1305,"currentLocation":"At London Bridge Platform 2","towards":"Morden via Bank","expectedArrival":"2017-10-09T08:26:28Z","timeToLive":"2017-10-09T08:26:28Z","modeName":"tube","timing":{"$type":"Tfl.Api.Presentation.Entities.PredictionTiming, Tfl.Api.Presentation.Entities","countdownServerAdjustment":"00:00:00","source":"0001-01-01T00:00:00","insert":"0001-01-01T00:00:00","read":"2017-10-09T08:04:36.369Z","sent":"2017-10-09T08:04:43Z","received":"0001-01-01T00:00:00"}}]
            
    //         const parsed = this._parseData(json);
    //         return Promise.resolve(parsed);
    // }
}
/* eslint-enable class-methods-use-this */

export default TfLDataAPIService;
