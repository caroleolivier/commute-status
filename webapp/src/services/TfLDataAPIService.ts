import TfLDataParser from './TfLDataParser';
import ArrivalTime from './ArrivalTime';

export interface TfLStationArrivalInput {
    stationId: string,
    directionId: string,
    lines: string[]
}

export interface FetchFacade {
    fetch(input: RequestInfo, init?: RequestInit): Promise<Response>;
}

export default class TfLDataAPIService {
    parser: TfLDataParser;
    fetchService: FetchFacade;

    constructor() {
        this.fetchService = { fetch: (params) => fetch(params) };
        this.parser = new TfLDataParser();
    }

    fetchArrivals(input: TfLStationArrivalInput): Promise<ArrivalTime[]> {
        const url = this._buildUrl(input.stationId, input.directionId, input.lines);
        return this.fetchService.fetch(url)
            .then(resp => resp.json())
            .then(json => this.parser.parse(json));
    }

    _buildUrl(stationId: string, directionId: string, lines: string[]): string {
        const linesStr = lines.join('%2C');
        return `https://api.tfl.gov.uk/Line/${linesStr}/Arrivals/${stationId}?direction=${directionId}`;
    }
}
