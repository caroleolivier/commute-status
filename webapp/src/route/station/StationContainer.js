import 'whatwg-fetch';
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import StationHeader from './StationHeader';
import ArrivalTimeTable from './ArrivalTimeTable';
import Loading from '../../common/Loading';
import ErrorMessage from '../../common/ErrorMessage';
import TfLDataAPIService from '../../services/TfLDataAPIService';
import NationalRailAPIService from '../../services/NationalRailAPIService';

import styles from './styles.scss';

const services = {
    tube: {
        get: () => new TfLDataAPIService()
    },
    bus: {
        get: () => new TfLDataAPIService()
    },
    train: {
        get: () => new NationalRailAPIService()
    }
};

const StationContainerState = {
    LOADING: {
        id: 0,
        getComponent: () => <Loading />
    },
    ERROR: {
        id: 1,
        getComponent: () => <ErrorMessage message="Unable to load data" />
    },
    LOADED: {
        id: 2,
        getComponent: container => <ArrivalTimeTable arrivals={container.state.expectedTimes} />
    }
};

class StationContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            state: StationContainerState.LOADING,
            expectedTimes: []
        };
    }

    componentDidMount() {
        const service = services[this.props.config.type].get();
        service.fetchArrivals(this.props.config)
            .then((arrivalTimes) => {
                /* lot of discussions around whether setState should be called here or not :s */
                /* eslint-disable react/no-did-mount-set-state */
                this.setState({
                    expectedTimes: arrivalTimes.slice(0, 4),
                    state: StationContainerState.LOADED
                });
                /* eslint-enable react/no-did-mount-set-state */
            })
            .catch((ex) => {
                console.log(ex);
                this.setState({
                    state: StationContainerState.ERROR
                });
            });
    }

    render() {
        return (
            <div className={styles.container}>
                <StationHeader
                    name={this.props.config.stationName}
                    type={this.props.config.type}
                />
                {this.state.state.getComponent(this)}
            </div>
        );
    }
}

StationContainer.propTypes = {
    config: PropTypes.shape({
        type: PropTypes.string,
        stationName: PropTypes.string,
        stationId: PropTypes.string,
        direction: PropTypes.string,
        directionId: PropTypes.string,
        lines: PropTypes.arrayOf(PropTypes.string)
    }).isRequired
};

export default StationContainer;

export { StationContainerState };
