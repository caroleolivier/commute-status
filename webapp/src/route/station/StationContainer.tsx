import 'whatwg-fetch';
import * as React from 'react';

import { IStationContainerConfig } from '../../config/configtype';
import { ArrivalTime } from '../../services/ArrivalTime';
import { StationHeader } from './StationHeader';
import { ArrivalTimeTable } from './arrivals/ArrivalTimeTable';
import { Loading } from '../../common/Loading';
import { ErrorMessage } from '../../common/ErrorMessage';
import { IDataAPIService } from '../../services/IDataAPIService';
import { TfLDataAPIService } from '../../services/TfLDataAPIService';
import { NationalRailAPIService } from '../../services/NationalRailAPIService';

import * as styles from './StationContainer.scss';

interface IDataApiServiceMapper {
    get(): IDataAPIService;
}

const services: Map<string, IDataApiServiceMapper> = new Map<string, IDataApiServiceMapper>([
    ['tube', { get: () => new TfLDataAPIService() }],
    ['bus', { get: () => new TfLDataAPIService() }],
    ['train', { get: () => new NationalRailAPIService() }]
]);

export const stationContainerStates: Map<string, IStationContainerStateItem> = new Map<string, IStationContainerStateItem>([
    ["LOADING", {
        id: 0,
        getComponent: () => <Loading />
    }],
    ["ERROR", {
        id: 1,
        getComponent: () => <ErrorMessage message="Unable to load data" />
    }],
    ["LOADED", {
        id: 2,
        getComponent: container => <ArrivalTimeTable arrivals={container.state.expectedTimes} />
    }]
]);

export interface IStationContainerProps {
    config: IStationContainerConfig;
}

export interface IStationContainerState {
    state: IStationContainerStateItem;
    expectedTimes: ArrivalTime[];
}

interface IStationContainerStateItem {
    id: number;
    getComponent(parent: StationContainer): JSX.Element;
}

export class StationContainer extends React.Component<IStationContainerProps, IStationContainerState> {
    constructor(props: IStationContainerProps) {
        super(props);
        const initialState: Readonly<IStationContainerState> = {
            state: stationContainerStates.get('LOADING'),
            expectedTimes: []
        };
        this.state = initialState;
    }

    componentDidMount(): void {
        const service: IDataAPIService = services.get(this.props.config.type).get();
        service.fetchArrivals(this.props.config)
            .then((arrivalTimes) => {
                /* lot of discussions around whether setState should be called here or not :s */
                this.setState({
                    expectedTimes: arrivalTimes.slice(0, 4),
                    state: stationContainerStates.get('LOADED')
                });
            })
            .catch((ex) => {
                console.log(ex);
                this.setState({
                    state: stationContainerStates.get('ERROR')
                });
            });
    }

    render(): JSX.Element {
        const stationContent: JSX.Element = this.state.state.getComponent(this);
        return (
            <div className={styles.container}>
                <StationHeader
                    stationName={this.props.config.stationName}
                    direction={this.props.config.travelDirection}
                    transportType={this.props.config.type}
                />
                {stationContent}
            </div>
        );
    }
}
