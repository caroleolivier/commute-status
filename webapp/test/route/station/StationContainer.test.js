/* eslint-env browser, jest */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';

import TfLDataAPIService from '../../../src/services/TfLDataAPIService.ts';

import StationContainer, { StationContainerState } from '../../../src/route/station/StationContainer';

describe('StationContainer component', () => {
    let stubService;

    const stationConfig = {
        type: 'bus',
        stationName: 'Waterloo',
        stationId: '1234DWWEF',
        direction: 'North',
        directionId: 'outbound',
        lines: ['jubilee']
    };
    let stationContainer;
    let fetchDataObj;
    let promise;

    beforeEach(() => {
        promise = new Promise((resolve, reject) => {
            fetchDataObj = { resolve, reject };
        });
        stubService = sinon.stub(TfLDataAPIService.prototype, 'fetchArrivals');
        stubService.returns(promise);
    });

    afterEach(() => {
        stubService.restore();
    });

    describe('Given component was mounted successfully', () => {
        beforeEach(() => {
            stationContainer = shallow(<StationContainer config={stationConfig} />);
        });

        test('it displays a loading state', () => {
            expect(stationContainer.instance().state.state).toEqual(StationContainerState.LOADING);
            const tree = toJson(stationContainer);
            expect(tree).toMatchSnapshot('loading');
        });

        test('it should send a request to load expected arrivals', () => {
            expect(stubService.calledWith(stationConfig)).toBeTruthy();
        });

        describe('Given the server request was successful', () => {
            test('it should load the expected time arrivals', () => {
                const arrivals = [
                    { lineName: '10', stationName: 'Waterloo', timeToStation: 60 },
                    { lineName: '10', stationName: 'Waterloo', timeToStation: 90 },
                    { lineName: '20', stationName: 'Waterloo', timeToStation: 30 }
                ];
                fetchDataObj.resolve(arrivals);

                return promise.then(() => {
                    // force wrapper to update
                    stationContainer.update();
                    expect(stationContainer.instance().state.state).toEqual(StationContainerState.LOADED);
                    const tree = toJson(stationContainer);
                    expect(tree).toMatchSnapshot('loaded');
                });
            });
        });

        describe('Given the server request failed', () => {
            test('it should display an error message', () => {
                fetchDataObj.reject('something wrong happened');

                return promise
                    .then(() => {
                        throw new Error('Should not be called');
                    })
                    .catch(() => {
                        // force wrapper to update
                        stationContainer.update();
                        expect(stationContainer.instance().state.state).toEqual(StationContainerState.ERROR);
                        const tree = toJson(stationContainer);
                        expect(tree).toMatchSnapshot('error');
                    });
            });
        });
    });
});
