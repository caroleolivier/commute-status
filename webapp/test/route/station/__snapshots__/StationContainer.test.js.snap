// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error 1`] = `
<div
  className="container"
>
  <StationHeader
    name="Waterloo"
    type="bus"
  />
  <ErrorMessage
    message="Unable to load data"
  />
</div>
`;

exports[`loaded 1`] = `
<div
  className="container"
>
  <StationHeader
    name="Waterloo"
    type="bus"
  />
  <ArrivalTimeTable
    arrivals={
      Array [
        Object {
          "lineName": "10",
          "stationName": "Waterloo",
          "timeToStation": 60,
        },
        Object {
          "lineName": "10",
          "stationName": "Waterloo",
          "timeToStation": 90,
        },
        Object {
          "lineName": "20",
          "stationName": "Waterloo",
          "timeToStation": 30,
        },
      ]
    }
  />
</div>
`;

exports[`loading 1`] = `
<div
  className="container"
>
  <StationHeader
    name="Waterloo"
    type="bus"
  />
  <Loading />
</div>
`;
